File: modules/reports/report_data_provider.js
/**
 * CerperStats Report Data Provider
 * 
 * Transforms raw session data into structured JSON for HTML report generation.
 * JavaScript port of the Python report_data_provider.py module.
 */

// =============================================================================
// Constants - Only truly fixed exceptions
// =============================================================================

// Only exception: These two DB categories map to the same header
const MERGED_CATEGORIES = {
    'Tratamiento de Resultados': 'PRUEBAS ESTADÍSTICAS APLICADAS A LOS RESULTADOS',
    'Atípicos': 'PRUEBAS ESTADÍSTICAS APLICADAS A LOS RESULTADOS',
};

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Map database category to display header.
 * Special cases handled via MERGED_CATEGORIES, otherwise uppercase.
 */
function getCategoryHeader(categoria) {
    if (!categoria) return 'OTROS';
    // Check for merged categories first
    if (MERGED_CATEGORIES[categoria]) {
        return MERGED_CATEGORIES[categoria];
    }
    // Default: just uppercase the DB category
    return categoria.toUpperCase();
}

/**
 * Escape HTML special characters
 */
function escapeHtml(value) {
    return String(value)
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#39;');
}

/**
 * Sanitize CSS class names
 */
function safeClassList(value) {
    return String(value || '')
        .split(/\s+/)
        .map(t => t.replace(/[^A-Za-z0-9_-]/g, '').trim())
        .filter(Boolean)
        .join(' ');
}

/**
 * Format a value for display in tables.
 * Uses value_mappings from config to apply labels and CSS classes.
 * @param {*} value - Value to format
 * @param {string} key - Column key (optional)
 * @param {Object} valueMappings - Value mappings from config
 */
function formatValue(value, key = null, valueMappings = {}) {
    // Null/undefined
    if (value === null || value === undefined) {
        return '<span class="df-value-null">-</span>';
    }

    // Check for value mappings (strings, booleans converted to string key)
    const mappingKey = typeof value === 'string' ? value : String(value);
    const mapping = valueMappings[mappingKey];

    if (mapping && typeof mapping === 'object') {
        const label = mapping.label != null ? mapping.label : mappingKey;
        const cls = mapping.class != null ? safeClassList(mapping.class) : '';
        const safeLabel = escapeHtml(label);
        const allClasses = safeClassList(`df-value-badge ${cls}`);
        return `<span class="${allClasses}">${safeLabel}</span>`;
    }

    // Column 'n' comes as string from Python, display as-is
    if (typeof value === 'string') {
        if (key && key.toLowerCase() === 'n') {
            return `<span class="df-value-number">${escapeHtml(value)}</span>`;
        }
        return `<span class="df-value-string">${escapeHtml(value)}</span>`;
    }

    // Numbers
    if (typeof value === 'number') {
        if (Number.isNaN(value)) return '<span class="df-value-null">-</span>';
        const formatted = Number.isInteger(value) ? String(value) : value.toFixed(4);
        return `<span class="df-value-number">${formatted}</span>`;
    }

    // Booleans (fallback if no mapping)
    if (typeof value === 'boolean') {
        return `<span class="df-value-string">${value ? 'Sí' : 'No'}</span>`;
    }

    return `<span class="df-value-string">${escapeHtml(String(value))}</span>`;
}

/**
 * Determine status (success/danger/neutral) from conclusion_status.
 * Uses the status provided by Python modules. Falls back to 'neutral' if not available.
 */
function getConclusionStatus(text, conclusionStatus = null) {
    // Use conclusion_status from Python if available and valid
    if (conclusionStatus && ['success', 'danger', 'neutral'].includes(conclusionStatus)) {
        return conclusionStatus;
    }
    
    // Default to neutral if no status provided
    return 'neutral';
}

/**
 * Transform snake_case key to Title Case label (fallback).
 * Example: 'p_value' → 'P Value', 'desviacion' → 'Desviacion'
 */
function keyToLabelFallback(key) {
    if (!key) return '';
    return key
        .replace(/_/g, ' ')
        .replace(/\b\w/g, c => c.toUpperCase());
}

/**
 * Get column label from config or fallback to Title Case.
 * @param {string} key - Column key
 * @param {Object} columnLabels - Column labels from config
 */
function getColumnLabel(key, columnLabels = {}) {
    const strKey = String(key || '');
    const dynamic = columnLabels[strKey];
    if (typeof dynamic === 'string' && dynamic.trim() !== '') return dynamic;
    return keyToLabelFallback(strKey);
}

/**
 * Priority order for columns (appear first)
 */
const PRIORITY_COLUMNS = ["parametro", "n", "prueba_normalidad", "prueba_homogeneidad", "prueba_tendencia"];

/**
 * Convert column keys to column objects with dynamic labels.
 * Applies priority ordering and uses column_labels from config.
 * @param {Set} keysSet - Set of column keys
 * @param {Object} columnLabels - Column labels from config
 */
function orderColumns(keysSet, columnLabels = {}) {
    const allKeys = new Set(keysSet);
    const orderedKeys = [];

    // Add priority columns first (in order)
    for (const key of PRIORITY_COLUMNS) {
        if (allKeys.has(key)) {
            orderedKeys.push(key);
            allKeys.delete(key);
        }
    }

    // Add remaining columns alphabetically
    orderedKeys.push(...Array.from(allKeys).sort());

    return orderedKeys.map(k => ({
        key: k,
        label: getColumnLabel(k, columnLabels)
    }));
}

/**
 * Sanitize color value for CSS (security)
 */
function sanitizeColor(value) {
    if (value === null || value === undefined) return null;
    const raw = String(value).trim();
    if (raw === '') return null;
    if (raw.toLowerCase() === 'transparent') return 'transparent';

    // #rgb, #rgba, #rrggbb, #rrggbbaa
    if (/^#([0-9a-fA-F]{3}|[0-9a-fA-F]{4}|[0-9a-fA-F]{6}|[0-9a-fA-F]{8})$/.test(raw)) {
        return raw;
    }

    // rgb()/rgba() with numeric channels only
    const m = raw.match(
        /^rgba?\(\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(\d{1,3})(?:\s*,\s*(0|1|0?\.\d+)\s*)?\)$/i
    );
    if (!m) return null;

    const r = Math.max(0, Math.min(255, Number(m[1])));
    const g = Math.max(0, Math.min(255, Number(m[2])));
    const b = Math.max(0, Math.min(255, Number(m[3])));
    const hasAlpha = typeof m[4] !== 'undefined';
    if (!hasAlpha) return `rgb(${r}, ${g}, ${b})`;

    const a = Math.max(0, Math.min(1, Number(m[4])));
    return `rgba(${r}, ${g}, ${b}, ${a})`;
}

/**
 * Sanitize angle value for CSS gradients
 */
function sanitizeAngle(value) {
    const n = Number(value);
    if (!Number.isFinite(n)) return null;
    return Math.max(0, Math.min(360, Math.round(n)));
}

/**
 * Build dynamic CSS rules from value_mappings (same logic as dataframe_renderer.js)
 * @param {Object} valueMappings - Value mappings from config
 * @returns {string} CSS rules
 */
function buildDynamicCssRules(valueMappings = {}) {
    const DYNAMIC_CLASS_PREFIX = 'df-';
    const classStyles = new Map();

    const mappings = valueMappings && typeof valueMappings === 'object'
        ? Object.values(valueMappings)
        : [];

    for (const mapping of mappings) {
        if (!mapping || typeof mapping !== 'object') continue;
        const style = mapping.style;
        if (!style || typeof style !== 'object') continue;

        const classTokens = safeClassList(mapping.class)
            .split(/\s+/)
            .filter(t => t.startsWith(DYNAMIC_CLASS_PREFIX));
        if (classTokens.length === 0) continue;

        const textColor = sanitizeColor(style.text_color ?? style.color ?? null);
        const bgFrom = sanitizeColor(style.bg_from ?? style.background_from ?? null);
        const bgTo = sanitizeColor(style.bg_to ?? style.background_to ?? null);
        const angle = sanitizeAngle(style.bg_angle ?? style.background_angle ?? null) ?? 135;

        for (const token of classTokens) {
            if (!token.startsWith(DYNAMIC_CLASS_PREFIX)) continue;
            classStyles.set(token, { textColor, bgFrom, bgTo, angle });
        }
    }

    const rules = [];
    for (const [token, style] of classStyles.entries()) {
        const decl = [];
        if (style.textColor) decl.push(`color: ${style.textColor};`);
        if (style.bgFrom && style.bgTo) {
            decl.push(`background: linear-gradient(${style.angle}deg, ${style.bgFrom} 0%, ${style.bgTo} 100%);`);
        } else if (style.bgFrom) {
            decl.push(`background-color: ${style.bgFrom};`);
        }
        if (decl.length === 0) continue;
        rules.push(`.${token} { ${decl.join(' ')} }`);
    }

    return rules.join('\n');
}

/**
 * Sanitize filename for Windows compatibility.
 * Removes invalid characters: < > : " / \ | ? * and control characters.
 */
function safeFilename(str) {
    if (!str) return 'unnamed';
    return str
        .replace(/\s+/g, '_')
        .replace(/[<>:"/\\|?*\x00-\x1F]/g, '-') // Remove Windows-invalid chars and control chars
        .replace(/-+/g, '-') // Collapse multiple dashes
        .replace(/^-|-$/g, ''); // Remove leading/trailing dashes
}

// =============================================================================
// Report Data Provider Class
// =============================================================================

class ReportDataProvider {
    /**
     * @param {Object} data - { session_id, session_info, results, graphs }
     * @param {Object} config - { group_by, include_graphs, include_tables, analyst_names, execution_date, logo_path, value_mappings, column_labels }
     */
    constructor(data, config) {
        this.sessionId = data.session_id;
        this.sessionInfo = data.session_info || {};
        this.results = data.results || [];
        this.graphs = data.graphs || [];
        this.config = config || {};

        // Index graphs by (catalog_id, nivel, analito)
        this.graphIndex = new Map();
        for (const g of this.graphs) {
            const key = `${g.catalog_id}-${g.nivel}-${g.analito || ''}`;
            this.graphIndex.set(key, g.grafico_data);
        }
    }

    _getGraph(catalogId, nivel, analito = null) {
        // Try specific match
        let g = this.graphIndex.get(`${catalogId}-${nivel}-${analito || ''}`);
        if (g) return g;
        // Try generic (no analito)
        return this.graphIndex.get(`${catalogId}-${nivel}-`);
    }

    /**
     * Main entry point. Returns array of report objects based on grouping config.
     */
    getReportData() {
        const groupBy = this.config.group_by || 'unified';

        switch (groupBy) {
            case 'by_analito_nivel':
                return this._generateByAnalitoNivel();
            case 'by_analito':
                return this._generateByAnalito();
            case 'by_nivel':
                return this._generateByNivel();
            default:
                return this._generateUnified();
        }
    }

    // =========================================================================
    // Generation Strategies
    // =========================================================================

    _generateUnified() {
        return [this._buildReportStructure(this.results, 'Unificado', 'unificado')];
    }

    _generateByNivel() {
        const byNivel = new Map();
        for (const r of this.results) {
            const nivel = r.nivel || 1;
            if (!byNivel.has(nivel)) byNivel.set(nivel, []);
            byNivel.get(nivel).push(r);
        }

        return Array.from(byNivel.entries())
            .sort((a, b) => a[0] - b[0])
            .map(([nivel, subset]) =>
                this._buildReportStructure(subset, `Nivel ${nivel}`, `nivel_${nivel}`, null, nivel)
            );
    }

    _generateByAnalito() {
        const byAnalito = new Map();
        for (const r of this.results) {
            const analito = r.analito;
            if (!analito) continue;
            if (!byAnalito.has(analito)) byAnalito.set(analito, []);
            byAnalito.get(analito).push(r);
        }

        return Array.from(byAnalito.entries())
            .sort((a, b) => a[0].localeCompare(b[0]))
            .map(([analito, subset]) =>
                this._buildReportStructure(subset, analito, safeFilename(analito), analito)
            );
    }

    _generateByAnalitoNivel() {
        const grouped = new Map();
        for (const r of this.results) {
            const analito = r.analito;
            const nivel = r.nivel || 1;
            if (!analito) continue;
            const key = `${analito}|${nivel}`;
            if (!grouped.has(key)) grouped.set(key, { analito, nivel, results: [] });
            grouped.get(key).results.push(r);
        }

        return Array.from(grouped.values())
            .sort((a, b) => a.analito.localeCompare(b.analito) || a.nivel - b.nivel)
            .map(({ analito, nivel, results }) =>
                this._buildReportStructure(
                    results,
                    `${analito} - Nivel ${nivel}`,
                    `${safeFilename(analito)}_nivel_${nivel}`,
                    analito,
                    nivel
                )
            );
    }

    // =========================================================================
    // Structure Builder
    // =========================================================================

    _buildReportStructure(resultsSubset, titleSuffix, filenameSuffix, filterAnalito = null, filterNivel = null) {
        // Generate dynamic CSS from value_mappings
        const dynamicCss = buildDynamicCssRules(this.config.value_mappings || {});

        return {
            filename: `informe_sesion_${this.sessionId}_${filenameSuffix}.pdf`,
            // Store metadata for main.js to extract
            _metadata: {
                analito: filterAnalito,
                nivel: filterNivel
            },
            data: {
                cover: this._buildCoverData(resultsSubset, titleSuffix),
                sections: this._buildSectionsData(resultsSubset, filterAnalito, filterNivel),
                // Pass logo_path to template
                logo_path: this.config.logo_path || null,
                // Pass tipo_analisis to template for table rendering logic
                tipo_analisis: this.sessionInfo.tipo_analisis || null,
                // Dynamic CSS for badge styles (from value_mappings)
                dynamic_css: dynamicCss
            }
        };
    }

    _buildCoverData(results, titleSuffix) {
        const info = this.sessionInfo;
        // Use execution_date from config (formatted as DD/MM/YYYY from pdf_config.js)
        // Fallback to current date if not provided
        const execDate = this.config.execution_date || new Date().toLocaleDateString('es-ES', {
            day: '2-digit',
            month: '2-digit',
            year: 'numeric'
        });

        // Participants (only if parametro == 'Analista')
        const participants = [];
        // Get analyst_names from config (array or null/undefined)
        const analystNames = this.config.analyst_names;
        // Verify that parametro is 'analista' (case-insensitive)
        const isAnalystParam = (info.parametro || '').toLowerCase() === 'analista';

        // Only add participants if:
        // 1. parametro is 'analista' AND
        // 2. analyst_names is provided (not null/undefined) AND
        // 3. analyst_names is a non-empty array
        if (isAnalystParam && Array.isArray(analystNames) && analystNames.length > 0) {
            analystNames.forEach((name, i) => {
                // Only add non-empty names
                if (name && String(name).trim()) {
                    participants.push({ index: `Analista ${i + 1}`, name: String(name).trim() });
                }
            });
        }

        // Tests list for cover
        const testsByCategory = new Map();
        const seenTests = new Set();

        for (const r of results) {
            const header = getCategoryHeader(r.categoria);
            const name = r.titulo || r.test_titulo || r.nombre_interno;
            const key = `${header}|${name}`;

            if (!seenTests.has(key)) {
                if (!testsByCategory.has(header)) testsByCategory.set(header, []);
                testsByCategory.get(header).push(name);
                seenTests.add(key);
            }
        }

        const testsList = [];
        // Use order of appearance in data (from testsByCategory map)
        const orderedHeaders = Array.from(testsByCategory.keys());

        for (const header of orderedHeaders) {
            testsList.push({
                header,
                tests: testsByCategory.get(header)
            });
        }

        return {
            title: 'INFORME ESTADÍSTICO',
            subtitle: titleSuffix,
            lab: info.lab_nombre || info.lab_key || '',
            expediente: info.expediente || '',
            fecha: execDate,
            ensayo: info.ensayo || '',
            metodo: info.metodo || '',
            producto: info.producto || '',
            unidad: info.unidad || '',
            parametro: info.parametro || '',
            participants,
            tests_list: testsList,
            signatures: {
                supervisor_name: info.supervisor_nombre || info.supervisor || '_________________',
                supervisor_role: 'Supervisor / Responsable de laboratorio'
            }
        };
    }

    _buildSectionsData(results, filterAnalito, filterNivel) {
        // Group: Header -> TestKey -> Nivel -> [results]
        const grouped = new Map();

        for (const r of results) {
            const header = getCategoryHeader(r.categoria);
            const testId = r.catalog_id;
            const testName = r.titulo || r.test_titulo || r.nombre_interno;
            const testDesc = r.descripcion || '';
            const testKey = `${testId}|${testName}|${testDesc}`;
            const nivel = r.nivel || 1;

            if (!grouped.has(header)) grouped.set(header, new Map());
            if (!grouped.get(header).has(testKey)) grouped.get(header).set(testKey, new Map());
            if (!grouped.get(header).get(testKey).has(nivel)) grouped.get(header).get(testKey).set(nivel, []);

            grouped.get(header).get(testKey).get(nivel).push(r);
        }

        // Build sections - use order of appearance in data
        const sections = [];
        const orderedHeaders = Array.from(grouped.keys());

        for (const header of orderedHeaders) {
            const testsMap = grouped.get(header);
            const testsData = [];

            // Sort by catalog_id
            const sortedTestKeys = Array.from(testsMap.keys()).sort((a, b) => {
                const idA = parseInt(a.split('|')[0]) || 0;
                const idB = parseInt(b.split('|')[0]) || 0;
                return idA - idB;
            });

            for (const testKey of sortedTestKeys) {
                const [testIdStr, testName, testDesc] = testKey.split('|');
                const testId = parseInt(testIdStr) || null;
                const nivelesMap = testsMap.get(testKey);
                const nivelesData = [];

                const sortedNiveles = Array.from(nivelesMap.keys()).sort((a, b) => a - b);

                for (const nivel of sortedNiveles) {
                    const rList = nivelesMap.get(nivel);
                    if (!rList.length) continue;

                    // Build table rows
                    const tableRows = [];
                    const colsSet = new Set();

                    for (const r of rList) {
                        let rawPc = r.resultado_pc;
                        if (typeof rawPc === 'string') {
                            try { rawPc = JSON.parse(rawPc); } catch { rawPc = []; }
                        }
                        if (!Array.isArray(rawPc)) rawPc = [];

                        for (const row of rawPc) {
                            const formatted = {};
                            for (const [k, v] of Object.entries(row)) {
                                formatted[k] = formatValue(v, k, this.config.value_mappings || {});
                                colsSet.add(k);
                            }
                            tableRows.push(formatted);
                        }
                    }

                    const columns = orderColumns(colsSet, this.config.column_labels || {});

                    // Build table only if include_tables is enabled
                    let table = null;
                    if (this.config.include_tables !== false && tableRows.length > 0) {
                        table = { rows: tableRows, columns };
                    }

                    // Graphs
                    const graphs = [];
                    if (this.config.include_graphs !== false) {
                        for (const r of rList) {
                            const analito = r.analito;
                            const gData = this._getGraph(testId, nivel, analito);
                            if (gData) {
                                graphs.push({
                                    data: gData.startsWith('data:') ? gData : `data:image/png;base64,${gData}`,
                                    label: rList.length > 1 ? analito : null
                                });
                            }
                        }
                    }

                    // Conclusion logic:
                    // - For multianalito: only show when grouped by analito (filterAnalito !== null)
                    // - For monoanalito: always show when it exists
                    let conclusion = null;
                    const tipoAnalisis = (this.sessionInfo.tipo_analisis || '').toLowerCase();
                    const isMultianalito = tipoAnalisis === 'multi' || tipoAnalisis === 'multianalito';
                    const groupBy = this.config.group_by || 'unified';
                    
                    // Show conclusions if:
                    // 1. It's monoanalito (always show), OR
                    // 2. It's multianalito AND we're filtering by analito (filterAnalito !== null) OR grouping by analito
                    const showConclusions = !isMultianalito || 
                        filterAnalito !== null || 
                        ['by_analito', 'by_analito_nivel'].includes(groupBy);
                    
                    if (showConclusions && rList.length > 0) {
                        // Try to find conclusion and conclusion_status from any result in the list
                        const resultWithConclusion = rList.find(r => r.conclusion) || rList[0];
                        const cText = resultWithConclusion?.conclusion;
                        const cStatus = resultWithConclusion?.conclusion_status || null;
                        if (cText) {
                            conclusion = {
                                text: cText,
                                status: getConclusionStatus(cText, cStatus)
                            };
                        }
                    }

                    nivelesData.push({
                        nivel,
                        table,
                        conclusion,
                        graphs
                    });
                }

                testsData.push({
                    name: testName,
                    description: testDesc,
                    niveles: nivelesData
                });
            }

            sections.push({
                header,
                tests: testsData
            });
        }

        return sections;
    }
}

module.exports = { ReportDataProvider };

File: modules/python/reports/report_data_provider.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
CerperStats Report Data Provider

Transforms raw session data into structured JSON for HTML report generation.
Replaces the direct PDF generation logic from monoanalito.py/multianalito.py.
"""

import json
from collections import defaultdict
from datetime import datetime

# =============================================================================
# Constants & Helpers
# =============================================================================

CATEGORY_HEADER_MAP = {
    'Tratamiento de Resultados': 'PRUEBAS ESTADÍSTICAS APLICADAS A LOS RESULTADOS',
    'Atípicos': 'PRUEBAS ESTADÍSTICAS APLICADAS A LOS RESULTADOS',
    'Veracidad': 'VERACIDAD',
    'Precisión': 'PRECISIÓN',
}

CATEGORY_ORDER = [
    'PRUEBAS ESTADÍSTICAS APLICADAS A LOS RESULTADOS',
    'VERACIDAD',
    'PRECISIÓN',
]

def format_value(value):
    """Format a value for display."""
    if value is None:
        return "-"
    if isinstance(value, bool):
        return "Sí" if value else "No"
    if isinstance(value, float):
        if value != value:  # NaN check
            return "-"
        return f"{value:.4f}" if abs(value) < 1000 else f"{value:.2f}"
    return str(value)

def get_conclusion_status(conclusion_text):
    """Determine status (success/danger/neutral) from conclusion text."""
    if not conclusion_text:
        return 'neutral'
    
    text = str(conclusion_text).lower()
    if any(w in text for w in ['cumple', 'normal', 'sí', 'aprobado', 'homogéneo']):
        return 'success'
    if any(w in text for w in ['no cumple', 'no normal', 'rechazado', 'no homogéneo']):
        return 'danger'
    return 'neutral'

def get_category_header(categoria):
    """Map database category to display header."""
    return CATEGORY_HEADER_MAP.get(categoria, categoria.upper() if categoria else 'OTROS')

# =============================================================================
# Main Data Provider Class
# =============================================================================

class ReportDataProvider:
    """Provides structured data for reports."""
    
    def __init__(self, data, config):
        self.session_id = data.get('session_id')
        self.session_info = data.get('session_info', {})
        self.results = data.get('results', [])
        self.graphs = data.get('graphs', [])
        self.config = config
        
        # Index graphs
        self._graph_index = {}
        for g in self.graphs:
            # Key: (catalog_id, nivel, analito) - analito can be None for mono
            key = (g.get('catalog_id'), g.get('nivel'), g.get('analito'))
            self._graph_index[key] = g.get('grafico_data')

    def _get_graph(self, catalog_id, nivel, analito=None):
        """Get graph data, trying specific analito first, then generic."""
        # Try specific match
        g = self._graph_index.get((catalog_id, nivel, analito))
        if g: return g
        
        # Try generic (None analito) if monoanalito
        return self._graph_index.get((catalog_id, nivel, None))

    def get_report_data(self):
        """
        Main entry point. Returns list of report objects based on grouping config.
        """
        group_by = self.config.get('group_by', 'unified')
        reports = []
        
        if group_by == 'by_analito_nivel':
            reports = self._generate_by_analito_nivel()
        elif group_by == 'by_analito':
            reports = self._generate_by_analito()
        elif group_by == 'by_nivel':
            reports = self._generate_by_nivel()
        else: # unified
            reports = self._generate_unified()
            
        return reports

    # =========================================================================
    # Generation Strategies
    # =========================================================================

    def _generate_unified(self):
        """All results in one report."""
        return [self._build_report_structure(
            self.results, 
            title_suffix="Unificado", 
            filename_suffix="unificado"
        )]

    def _generate_by_nivel(self):
        """One report per nivel."""
        by_nivel = defaultdict(list)
        for r in self.results:
            by_nivel[r.get('nivel', 1)].append(r)
            
        reports = []
        for nivel, subset in by_nivel.items():
            reports.append(self._build_report_structure(
                subset,
                title_suffix=f"Nivel {nivel}",
                filename_suffix=f"nivel_{nivel}",
                filter_nivel=nivel
            ))
        return reports

    def _generate_by_analito(self):
        """One report per analito."""
        by_analito = defaultdict(list)
        for r in self.results:
            analito = r.get('analito')
            if analito:
                by_analito[analito].append(r)
                
        reports = []
        for analito, subset in by_analito.items():
            safe_analito = analito.replace(' ', '_').replace('/', '-')
            reports.append(self._build_report_structure(
                subset,
                title_suffix=analito,
                filename_suffix=safe_analito,
                filter_analito=analito
            ))
        return reports

    def _generate_by_analito_nivel(self):
        """One report per analito + nivel."""
        grouped = defaultdict(list)
        for r in self.results:
            analito = r.get('analito')
            nivel = r.get('nivel', 1)
            if analito:
                grouped[(analito, nivel)].append(r)
                
        reports = []
        for (analito, nivel), subset in grouped.items():
            safe_analito = analito.replace(' ', '_').replace('/', '-')
            reports.append(self._build_report_structure(
                subset,
                title_suffix=f"{analito} - Nivel {nivel}",
                filename_suffix=f"{safe_analito}_nivel_{nivel}",
                filter_analito=analito,
                filter_nivel=nivel
            ))
        return reports

    # =========================================================================
    # Structure Builder
    # =========================================================================

    def _build_report_structure(self, results_subset, title_suffix="", filename_suffix="", filter_analito=None, filter_nivel=None):
        """Build the complete JSON structure for a single report."""
        
        # 1. Prepare Cover Data
        cover_data = self._build_cover_data(results_subset, title_suffix)
        
        # 2. Prepare Sections Data
        sections_data = self._build_sections_data(results_subset, filter_analito, filter_nivel)
        
        return {
            "filename": f"informe_sesion_{self.session_id}_{filename_suffix}.pdf",
            "data": {
                "cover": cover_data,
                "sections": sections_data
            }
        }

    def _build_cover_data(self, results, title_suffix):
        info = self.session_info
        exec_date = self.config.get('execution_date') or datetime.now().strftime('%d/%m/%Y')
        
        # Participants
        participants = []
        analyst_names = self.config.get('analyst_names', [])
        is_analyst_param = (info.get('parametro') or '').lower() == 'analista'
        
        if is_analyst_param and analyst_names:
            for i, name in enumerate(analyst_names, 1):
                participants.append({"index": f"Analista {i}", "name": name})

        # Tests List for Cover
        test_names_by_category = defaultdict(list)
        seen_tests = set()
        for r in results:
            cat_header = get_category_header(r.get('categoria'))
            name = r.get('titulo') or r.get('test_titulo') or r.get('nombre_interno')
            
            if (cat_header, name) not in seen_tests:
                test_names_by_category[cat_header].append(name)
                seen_tests.add((cat_header, name))
        
        tests_list = []
        for header in CATEGORY_ORDER + [k for k in test_names_by_category.keys() if k not in CATEGORY_ORDER]:
            if header in test_names_by_category:
                tests_list.append({
                    "header": header,
                    "tests": test_names_by_category[header]
                })

        return {
            "title": "INFORME ESTADÍSTICO",
            "subtitle": title_suffix, # Can be used in template if needed
            "lab": info.get('lab_nombre') or info.get('lab_key', ''),
            "expediente": info.get('expediente', ''),
            "fecha": exec_date,
            "ensayo": info.get('ensayo', ''),
            "metodo": info.get('metodo', ''),
            "producto": info.get('producto', ''),
            "unidad": info.get('unidad', ''),
            "parametro": info.get('parametro', ''),
            "participants": participants,
            "tests_list": tests_list,
            "logo": self.config.get('logo_path_url'), # Needs special handling for file:// probably
            "signatures": {
                "supervisor_name": info.get('supervisor_nombre') or info.get('supervisor') or '_________________',
                "supervisor_role": "Supervisor / Responsable de laboratorio"
            }
        }

    def _build_sections_data(self, results, filter_analito, filter_nivel):
        """Group results into sections -> tests -> niveles."""
        
        # Group by Header -> Test(CatalogID) -> Nivel
        grouped = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))
        
        for r in results:
            cat_header = get_category_header(r.get('categoria'))
            
            test_id = r.get('catalog_id')
            test_name = r.get('titulo') or r.get('test_titulo') or r.get('nombre_interno')
            test_desc = r.get('descripcion', '')
            
            # Key for test uniqueness
            test_key = (test_id, test_name, test_desc)
            
            nivel = r.get('nivel', 1)
            
            grouped[cat_header][test_key][nivel].append(r)
            
        # Build output structure
        sections = []
        
        # Order categories
        headers = [h for h in CATEGORY_ORDER if h in grouped] + [h for h in grouped if h not in CATEGORY_ORDER]
        
        for header in headers:
            tests_data = []
            test_keys_sorted = sorted(grouped[header].keys(), key=lambda x: x[0] or 0) # Sort by catalog_id
            
            for test_key in test_keys_sorted:
                test_id, test_name, test_desc = test_key
                niveles_dict = grouped[header][test_key]
                
                niveles_data = []
                sorted_niveles = sorted(niveles_dict.keys())
                
                for nivel in sorted_niveles:
                    r_list = niveles_dict[nivel]
                    if not r_list: continue
                    
                    # Prepare Table
                    table_rows = []
                    cols = set()
                    
                    for r in r_list:
                        # Parse JSONB
                        raw_pc = r.get('resultado_pc')
                        if isinstance(raw_pc, str):
                            try:
                                raw_pc = json.loads(raw_pc)
                            except:
                                raw_pc = []
                        if not isinstance(raw_pc, list): raw_pc = []
                        
                        analito_label = r.get('analito')
                        
                        for row in raw_pc:
                            formatted = {k: format_value(v) for k,v in row.items()}
                            if analito_label and len(r_list) > 1: # If multianalito mixed
                                formatted['_analito'] = analito_label # Internal key? Or display
                                # Actually, user wants tables concatenated usually
                            table_rows.append(formatted)
                            cols.update(formatted.keys())
                            
                    # Table Columns Logic
                    columns = self._order_columns(cols)
                    
                    # Graphs
                    graphs = []
                    include_graphs = self.config.get('include_graphs', True)
                    if include_graphs:
                        # Collect graphs for this test+nivel
                        # If multianalito, we might have multiple
                        for r in r_list:
                             analito = r.get('analito')
                             g_data = self._get_graph(test_id, nivel, analito)
                             if g_data:
                                 graphs.append({
                                     "data": g_data if g_data.startswith('data:') else f"data:image/png;base64,{g_data}",
                                     "label": analito if len(r_list) > 1 else None
                                 })

                    # Conclusion
                    # Usually just one conclusion per analito.
                    # If mixed (unified), usually we pick one or show multiple?
                    # The old logic:
                    # by_analito -> show conclusion
                    # unified -> NO conclusion (because it's mixed)
                    conclusion = None
                    show_conclusions = (filter_analito is not None) or (self.config.get('group_by') in ['by_analito', 'by_analito_nivel'])
                    
                    if show_conclusions and r_list:
                         c_text = r_list[0].get('conclusion')
                         if c_text:
                             conclusion = {
                                 "text": c_text,
                                 "status": get_conclusion_status(c_text)
                             }

                    niveles_data.append({
                        "nivel": nivel if len(sorted_niveles) > 1 or filter_nivel is None else None, # Hide nivel title if filtered by single level?
                        # Actually old logic always shows "- Nivel X"
                        "nivel_label": f"Nivel {nivel}", 
                        "table": {"rows": table_rows, "columns": columns},
                        "conclusion": conclusion,
                        "graphs": graphs
                    })
                    
                tests_data.append({
                    "name": test_name,
                    "description": test_desc,
                    "niveles": niveles_data
                })
                
            sections.append({
                "header": header,
                "tests": tests_data
            })
            
        return sections

    def _order_columns(self, keys):
        """Order columns logically."""
        priority = ['analito', 'parametro', 'n', 'media', 'desviacion', 'cv', 'asimetria', 'curtosis', 'p_value', 'estadistico', 'normalidad', 'homogeneidad']
        ordered = []
        keys = set(keys)
        
        for k in priority:
            if k in keys:
                ordered.append(k)
                keys.remove(k)
        
        ordered.extend(sorted(list(keys)))
        
        # Map to label objects
        labels = {
            'n': 'n',
            'media': 'Media',
            'desviacion': 'Desv. Est.',
            'asimetria': 'Asimetría',
            'curtosis': 'Curtosis',
            'p_value': 'P-Value',
            'normalidad': 'Normalidad',
            'analito': 'Analito',
            'parametro': 'Parámetro',
             # Add other mappings...
        }
        
        return [{"key": k, "label": labels.get(k, k.replace('_', ' ').title())} for k in ordered]
